### ISSUE TITLE
No peer certificate in context when using netty4.

### ISSUE THREAD

**Issue by koshelev** [twitter/finagle#600]
When using `finagle-netty4` from the `develop` branch, peer certificate is not available in the context after successful TLS handshake.

### Expected behavior

If server is configured to use mutual TLS authentication and TLS handshake was successful, than service can query client's certificate using `com.twitter.finagle.transport.Transport.peerCertificate`.

### Actual behavior

`Transport.peerCertificate` returns `None`. 
[ChannelTransport.peerCertificate](https://github.com/twitter/finagle/blob/173a090a26745afaddbb686f0f04553b3cd1ebb5/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/transport/ChannelTransport.scala#L193) is initialized before handshake is finished.

### Steps to reproduce the behavior

```scala
package com.twitter.finagle.http

import java.net.InetSocketAddress

import com.twitter.finagle.{Address, Service, Http => FinagleHttp}
import com.twitter.finagle.ssl.client.SslClientConfiguration
import com.twitter.finagle.ssl.{ClientAuth, Engine}
import com.twitter.finagle.ssl.server.SslServerConfiguration
import com.twitter.finagle.stats.NullStatsReceiver
import com.twitter.finagle.transport.Transport
import com.twitter.util.{Await, Future}
import org.scalatest.FunSuite
import io.netty.handler.ssl.SslContextBuilder
import io.netty.handler.ssl.util.SelfSignedCertificate

class SslHandshakeEndToEndTest extends FunSuite {

  test("Peer certificate is available") {

    val service = new Service[Request, Response] {
      def apply(request: Request): Future[Response] = {
        val response = Response( Transport.peerCertificate.fold(Status.NotFound)(_ => Status.Ok) )
        Future.value(response)
      }
    }

    val serverCert = new SelfSignedCertificate()
    val clientCert = new SelfSignedCertificate()
    val allocator = io.netty.buffer.UnpooledByteBufAllocator.DEFAULT
    val server = FinagleHttp.server
      .configured(FinagleHttp.Netty4Impl)
      .withTransport.tls(SslServerConfiguration(clientAuth = ClientAuth.Needed), (config: SslServerConfiguration) => {
      val context = SslContextBuilder.forServer(serverCert.key(), serverCert.cert()).trustManager(clientCert.cert()).build()
      val engine = context.newEngine(allocator)
      engine.setNeedClientAuth(true)
      new Engine(engine)
    }).withStatsReceiver(NullStatsReceiver).serve("localhost:*", service)

    val addr = server.boundAddress.asInstanceOf[InetSocketAddress]
    val client = FinagleHttp.client
      .withTransport.tls(SslClientConfiguration(), (address: Address, config: SslClientConfiguration) => {
      val ctx = SslContextBuilder.forClient.keyManager(clientCert.key(), clientCert.cert()).trustManager(serverCert.cert()).build()
      new Engine(ctx.newEngine(allocator))
    }).withStatsReceiver(NullStatsReceiver).newService(s"${addr.getHostName}:${addr.getPort}", "client")

    assert( Await.result( client( Request(Method.Get, "/")) ).status == Status.Ok )
  }
}
```

**Comment by vkostyukov** [comment]
Thanks @koshelev! As I mentioned on Gitter, I like your idea about changing `peerCertificate` on N4 transport from `val` to `def`. Any chance you want to open a PR?

**Comment by koshelev** [comment]
@vkostyukov I'm afraid the solution with changing `val` to `def` will not fix the problem completely. [This](https://github.com/twitter/finagle/blob/9fa0dafb59080f5e600e0d00f3012a8ac66a5216/finagle-core/src/main/scala/com/twitter/finagle/server/StackServer.scala#L282) will still be broken. :(

**Comment by vkostyukov** [comment]
Oh I see. So we won't propagate that further. That's a bummer. Seems like having that connect delay handler on a server-side is our only option. What do you think @koshelev?

**Comment by koshelev** [comment]
Seems so. On the bright side delaying will have another nice side effect - service factory will not instantiate the service if handshake fails. Kind of DDoS protection.
@vkostyukov should adding `SslConnectHandler` to `Netty4ServerSslHandler` fix the problem or some more involved solution needed?

**Comment by vkostyukov** [comment]
/I think/ it should just work even though it was written with the client-side use case in mind. All the things in `SslHandler` and in the pipeline (at that stage) should be symmetric for client and server. I'd say let's try it. @koshelev Do you want to volunteer and open a PR?

**Comment by koshelev** [comment]
I will try it tomorrow (it is already midnight here in Berlin). I will open a PR in case it work. If not - I have some time to spend on it.

**Comment by vkostyukov** [comment]
Perfect! Thanks a lot @koshelev!

**Comment by koshelev** [comment]
@vkostyukov Seems like we need different implementation of `SslConnectHandler` for server. The problem is `ChannelInboundHandler` receives no connect promise, so we have nothing to delay. I would suggest to change `Netty4Listener` to add only raw initializer to a pipeline and move the initialization code starting from [here](https://github.com/twitter/finagle/blob/2e08384a7cf17821cbd6cf1d0175808bc045daa1/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/Netty4Listener.scala#L172) into new class `ServerSslConnectHandler`. It will add marshalling, framing and bridge if handshake succeeds. What do you think?

**Comment by vkostyukov** [comment]
You're right. There is no notion of connect-promise on the server-side. Although, /I think/ we can delay either `channelInit` or `channelActive` on a child pipeline in order to make sure this [server bridge](https://github.com/twitter/finagle/blob/develop/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/Netty4Listener.scala#L179) (essentially, a channel transport) thing is created only when the handshake is complete.

It seems like we do [a similar thing for Netty 3](https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/netty3/ssl/SslConnectHandler.scala#L18-L46). There we are delaying the `channelConnected` event, which is equivalent to `channelActive` in Netty 4.

**Comment by koshelev** [comment]
@vkostyukov Opened PR. Tried solution you proposed. The problem with it is the way `Netty4Listener` [initialises a pipeline](https://github.com/twitter/finagle/blob/2e08384a7cf17821cbd6cf1d0175808bc045daa1/finagle-netty4/src/main/scala/com/twitter/finagle/netty4/Netty4Listener.scala#L160-L183) - a channel initializer adds more channel initializers. New initializers will be [fired immediately](https://github.com/netty/netty/blob/eb7f751ba519cbcab47d640cd18757f09d077b55/transport/src/main/java/io/netty/channel/ChannelInitializer.java#L100)

**Comment by vkostyukov** [comment]
Fixed in #603.
